### ISSUE TITLE
Should DefaultDist be exposed as a user facing distribution?

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#14280]
Should `DefaultDist` be exposed as a user facing distribution? It would act as a kind of "NOP distribution", effectively distributing no computation but allowing idioms such as:

```chapel
use LayoutCS;
use BlockDist;

config type LayoutType = DefaultDist;
const dom = {0..16, 0..16};
const sparseDom: sparse subdomain(dom) dmapped LayoutType();
```

Even the alternative (which does not currently work, see #14279) is less elegant in most cases:

```chapel
use LayoutCS;
use BlockDist;

config param val = 0;

var parentDom = {1..2, 1..2};
var myDom: if val == 0 then
                sparse subdomain(parentDom) dmapped CS()
               else
                sparse subdomain(parentDom);
```

---

Many of the sparse tests that are currently leaking (see `users/engin/sparse_bulk/sparseAdd2D.chpl` as one example)  rely on using `DefaultDist` as a means of specifying "use the default layout", creating a distribution that is effectively a NOP.

The `DefaultDist` _type_ is also used by some distribution initializers (i.e, `Block`) as a way of specifying a sane default for sparse layouts.

---

As it stands today, `DefaultDist` is not intended to be used at the user level.

```chapel

const dom1 = {1..32};
const dom2: domain(1) dmapped DefaultDist() = dom1;
```

This will result in the leak of one `DefaultDist` (48 bytes). Upon inspecting the code for `DefaultDist` in `modules/internal/DefaultRectangular.chpl`, the reason becomes clear: `DefaultDist` is written as a singleton class that is not intended to be constructed multiple times (whether via `dmapped` or `new`).

Instead of being a "proper" singleton class, `DefaultDist` seems to rely on the assumption that it will never have an instance constructed publicly.

---

**What should be done about `DefaultDist`?** Here are a few options debated by @e-kayrakli and I.

**1A)** Don't expose DefaultDist as a public facing type in any capacity whatsoever. This would preclude it from being used in initializers and error messages (where it can currently appear).

**1B)** Refactor DefaultDist so that it is a "true singleton". Any new instance of `DefaultDist` that is created would be a handle to the singleton. This would allow `DefaultDist` to be used as a "NOP distribution". 

**1C)** Create a new, public facing distribution intended to be used as a "NOP distribution". Effectively `DefaultDist`, but with a more meaningful name. (Renaming `DefaultDist` would also work).

**Comment by e-kayrakli** [comment]
In the short term, I think a user-facing means for saying "use default distribution/layout" explicitly should be added and we should accept that `DefaultDist` is an implementation quirk that is just exposed to the user but shouldn't be used.

For layouts, I think we can add `COO`, so that we can do something like:
```chapel
var sparseDom: sparse subdomain(parentDom) dmapped COO();
```

For distributions, we can consider `NoDist` so that:
```chapel
var myDom = problemSpace dmapped NoDist();
```

would be equivalent to

```chapel
var myDom = problemSpace;
```

I think this would make some idioms less cumbersome, and from a more philosophical standpoint I think we should be able to explicitly specify the default behavior using user-facing means, rather than omitting clauses (`dmapped`) or using contrived means.

**Comment by bradcray** [comment]
I agree that the default distribution(s) should be exposed to the programmer, but not via the name `DefaultDist`.  I think that, as with many things, coming up with a good name is the big challenge.  I'm not convinced that `NoDist()` is the right name because there are many domain maps that don't distribute (and hopefully more over time), so it doesn't really distinguish well.

**Comment by e-kayrakli** [comment]
I am not crazy about `NoDist` myself either

It should be at the very least "class-like". So we can do `dmapped Foo()`. I was thinking about `dmapped none` or something like that but it may not help making things analogous.

I also thought about `dmapped Local()`, which kind of makes sense in the declaration. But at the same time can create a confusion: "does this mean the domain is always local?"
